### PR TITLE
Add top comments to articles feed

### PR DESCRIPTION
**Pull request by joshpuetz**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This PR adds 2 highest "top comments" (comments with a score higher than 10) to the feed used to populated the home page. These comments are cached for quicker retrieval when generating the home page feed.

## Related Tickets & Documents
Pitch document: https://docs.google.com/document/d/1m_jKPM9xhzFlR1wp9pVcihWFlNOigEHqsNWLEGYZb_E/edit#bookmark=id.k8ozjcscezgz

## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [ ] no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

McsWCjgBIZvwagK8b5

### CODE REVIEW

**Comment by benhalpern** [comment]
> only including comments for some article types (#ama, #watercooler), and caching these for better performance.

IMO this isn't _necessarily_ needed. Just allowing _top_ comments for any post type might be enough but we could definitely segment it (#discuss, #watercooler)...

Either way, cool to see this happening!

I imagine we want to limit this to _root_ comments (where `ancestry: nil`) eh? And are we showing _all_ comments on a post with a minimal score count, or just the one top one?

Maybe we've addressed this elsewhere I didn't see but I'm not sure if those details have been totally fleshed out yet so asking here.

**Comment by joshpuetz** [comment]
Great points @benhalpern!

> IMO this isn't _necessarily_ needed. Just allowing _top_ comments for any post type might be enough but we could definitely segment it (#discuss, #watercooler)...

It might be overkill, but my thought is unless we want to show top comments for every article card in the feed (which I think might be overkill), I'd like the logic to select which articles get top comments to be on the server side and not client side.

> I imagine we want to limit this to _root_ comments (where `ancestry: nil`) eh? And are we showing _all_ comments on a post with a minimal score count, or just the one top one?

That's an interesting thought: I ran a query (https://dev.to/internal/blazer/queries/15-highly-rated-comments-root-vs-child) and a highly rated comment is about 3 times more likely to be a root comment versus leaf: I'm not sure if that justifies filtering out leaf comments or not. I'll add it in for now!

We're only showing two individual comments with scores above 10, not all comments with scores above 10.

**Comment by joshpuetz** [comment]
I'm having an issue with the `bullet` gem: it both insists I add eager loading and remove it!

```
15:30:45 web.1       | GET /stories/feed/?page=1
15:30:45 web.1       | USE eager loading detected
15:30:45 web.1       |   Article => [:top_comments]
15:30:45 web.1       |   Add to your query: .includes([:top_comments])
```

Ok, expected, so we'll do some eager loading of top comments were we load up articles for the feed...

```
15:28:28 web.1       | GET /?i=i
15:28:28 web.1       | AVOID eager loading detected
15:28:28 web.1       |   Article => [:top_comments]
15:28:28 web.1       |   Remove from your query: .includes([:top_comments])
```

Any thoughts? Seems as though there's a conflict of interests between what's the best preloading course of action for the logged out homepage versus the logged in homepage.

File: app/models/article.rb [hunk 1]
```diff
@@ -467,6 +479,7 @@ def detect_human_language
   def async_score_calc
     return if !published? || destroyed?
 
+    Rails.cache.delete(top_comments_cache_name)
```
Comment (joshpuetz) [inline app/models/article.rb:482]: I'm conflicted about the location of this cache busting action, here are my options:

- bust the cache here in `articles` before recalculating article score (which happens when comments change)
- bust the cache in the async action (but I don't know if that's feasible: are sidekiq jobs in a separate dyno? Can they see the Rails cache?)
- bust somewhere in `comments` model since that's what we are really responding to (but: that would require the load of the article object in a comment only action..seems wrong!)

Comment (nickytonline) [reply app/models/article.rb:491]: The last option seems wrong as you mentioned. I'm probably missing some context, but what is the advantage of busting the cache in the async action vs. in the article model?

File: spec/requests/stories/feeds_spec.rb [hunk 2]
```diff
@@ -114,6 +114,17 @@
       end
     end
 
+    context "when there are highly rated comments" do
+      let(:comment) { create(:comment, score: 20) }
+      let(:article) { comment.commentable }
```
Comment (nickytonline) [inline spec/requests/stories/feeds_spec.rb:119]: What does the `commentable` property represent? Is this related to the "Hide" comment feature?

File: app/models/article.rb [hunk 3]
```diff
@@ -29,6 +29,7 @@ class Article < ApplicationRecord
   counter_culture :organization
 
   has_many :comments, as: :commentable, inverse_of: :commentable
+  has_many :top_comments, -> { where("comments.score > ? and ancestry is null", 10).order("comments.score DESC").limit(2) }, as: :commentable, inverse_of: :commentable, class_name: "Comment"
```
[the remark this answers is not in the document]
Comment (joshpuetz) [reply app/models/article.rb:32]: The later: it grabs only comments that are direct replies to the original story (top level comments)

File: app/models/article.rb (hunk 1 above)
Comment (joshpuetz) [reply app/models/article.rb:491]: Yeah: async is generally reserved for operations that could take a while, and/or we don't want to block the user while the action happens. Cache clearing will be very quick, so I don't think doing I asynchronously is advantageous here.

File: config/environments/development.rb [hunk 4]
```diff
@@ -126,6 +127,7 @@
     Bullet.add_whitelist(type: :unused_eager_loading, class_name: "ApiSecret", association: :user)
     # acts-as-taggable-on has super weird eager loading problems: <https://github.com/mbleigh/acts-as-taggable-on/issues/91>
     Bullet.add_whitelist(type: :n_plus_one_query, class_name: "ActsAsTaggableOn::Tagging", association: :tag)
+    Bullet.add_whitelist(type: :unused_eager_loading, class_name: "Article", association: :top_comments)
```
Comment (joshpuetz) [inline config/environments/development.rb:130]: Whitelisted the association here as Bullet was flip flopping between telling me to include and not included. Seems a riff on this problem: https://github.com/flyerhzm/bullet/issues/147

File: config/environments/development.rb [hunk 5]
```diff
@@ -1,3 +1,4 @@
+# rubocop:disable Metrics/BlockLength
```
Comment (joshpuetz) [inline config/environments/development.rb:1]: nCrggtBjL3p1C

File: app/models/article.rb [hunk 6]
```diff
@@ -417,6 +419,16 @@ def update_score
                    spaminess_rating: BlackBox.calculate_spaminess(self))
   end
 
+  def top_comments_cache_name
```
Comment (rhymes) [inline app/models/article.rb:422]: I think this should be private

File: spec/models/article_spec.rb [hunk 7]
```diff
@@ -845,6 +845,38 @@ def build_and_validate_article(*args)
     end
   end
 
+  describe "#top_comments" do
```
Comment (rhymes) [inline spec/models/article_spec.rb:848]: we should add a test for `cached_top_comments` if we decide to keep it

**Comment by joshpuetz** [comment]
> I'm having an issue with the `bullet` gem: it both insists I add eager loading and remove it!
> 
> ```
> 15:30:45 web.1       | GET /stories/feed/?page=1
> 15:30:45 web.1       | USE eager loading detected
> 15:30:45 web.1       |   Article => [:top_comments]
> 15:30:45 web.1       |   Add to your query: .includes([:top_comments])
> ```
> 
> Ok, expected, so we'll do some eager loading of top comments were we load up articles for the feed...
> 
> ```
> 15:28:28 web.1       | GET /?i=i
> 15:28:28 web.1       | AVOID eager loading detected
> 15:28:28 web.1       |   Article => [:top_comments]
> 15:28:28 web.1       |   Remove from your query: .includes([:top_comments])
> ```
> 
> Any thoughts? Seems as though there's a conflict of interests between what's the best preloading course of action for the logged out homepage versus the logged in homepage. Perhaps it's an issue of `top_comments` currently being referenced in the json generated to produce a logged in user feed, but it's not referenced in the logged out user/non-preact feed yet.

I found an issue referenced in the Bullet gem archives that matches the behavior I'm seeing (https://github.com/flyerhzm/bullet/issues/147): they really didn't fix it per se, so I've whitelisted the `articles.top_comments` association in Bullet's config.

File: spec/requests/stories/feeds_spec.rb (hunk 2 above)
Comment (joshpuetz) [reply spec/requests/stories/feeds_spec.rb:119]: It's the link back to `article` on the `comment`. A comment can be attached to several different classes (`article`, `podcast_episode`, or `user`, that's why it's generalized here

File: app/models/article.rb [hunk 8]
```diff
@@ -417,6 +419,16 @@ def update_score
...
+    "article-#{id}-#{last_comment_at&.rfc3339}/top_comments"
+  end
+
+  def cached_top_comments
+    Rails.cache.fetch(top_comments_cache_name, expires_in: 24.hours) do
```
[the remark this answers is not in the document]
Comment (joshpuetz) [reply app/models/article.rb:431]: No, that all makes sense: I have **no idea** how this will perform TBH. Here's an `explain analyze` of the query to get top comments:

```
Sort  (cost=444.67..444.67 rows=12 width=944) (actual time=3.646..3.646 rows=2 loops=1)
  Sort Key: comments.score
  Sort Method: quicksort  Memory: 28kB
  ->  Bitmap Heap Scan on comments  (cost=402.83..444.62 rows=12 width=944) (actual time=3.614..3.625 rows=2 loops=1)
        Recheck Cond: ((commentable_id = ANY ('{303199,302536,300600,302629,302747,302603,302462,302630,303213,302601,302461,302816,303266,302688,303222,302483,302634,303080,301184,301941,302616,302642,303359,302328,302769,302454,302751,302480,302740,302679,302622,302495,302833,302785,302826}'::integer[])) AND ((commentable_type)::text = 'Article'::text) AND (score > 10))
        Filter: (ancestry IS NULL)
        Heap Blocks: exact=2
        ->  BitmapAnd  (cost=402.83..402.83 rows=21 width=0) (actual time=3.586..3.586 rows=0 loops=1)
              ->  Bitmap Index Scan on index_comments_on_commentable_id_and_commentable_type  (cost=0.00..74.08 rows=547 width=0) (actual time=0.167..0.167 rows=187 loops=1)
                    Index Cond: ((commentable_id = ANY ('{303199,302536,300600,302629,302747,302603,302462,302630,303213,302601,302461,302816,303266,302688,303222,302483,302634,303080,301184,301941,302616,302642,303359,302328,302769,302454,302751,302480,302740,302679,302622,302495,302833,302785,302826}'::integer[])) AND ((commentable_type)::text = 'Article'::text))
              ->  Bitmap Index Scan on index_comments_on_score  (cost=0.00..328.70 rows=15076 width=0) (actual time=3.351..3.351 rows=19210 loops=1)
                    Index Cond: (score > 10)
Planning Time: 0.878 ms
Execution Time: 3.703 ms
```

It doesn't look too bad? I think the best course of action right now would be to leave out the Rails caching of top comments, and add it in if we see performance is lagging.

[the remark this answers is not in the document]
Comment (rhymes) [reply app/models/article.rb]: Yeah, 3ms is probably less than the roundtrip to Redis :-) Let's test it out in production :D

**Comment by joshpuetz** [comment]
> Looks fine to me, though removing the limit might result with too many comments loaded for some articles? Don't know how realistic is that scenario though 🤷‍♂️

I removed the limit for 2 reasons:

- I feel like the frontend or controller should be making the decision about how many top comments to include, not the join condition
- having a `limit` clause was applying to the entire select of comments, not just the association as part of the include. As a result: only a max of two top comments **across all the articles** were being loaded 😱

File: app/models/article.rb [hunk 9]
```diff
@@ -29,6 +29,11 @@ class Article < ApplicationRecord
   counter_culture :organization
 
   has_many :comments, as: :commentable, inverse_of: :commentable
+  has_many :top_comments,
```
Comment (mstruve) [inline app/models/article.rb:32]: 🔥

**Comment by rhymes** [comment]
@joshpuetz 

> > Looks fine to me, though removing the limit might result with too many comments loaded for some articles? Don't know how realistic is that scenario though 🤷‍♂️
> 
> I removed the limit for 2 reasons:
> 
>     * I feel like the frontend or controller should be making the decision about how many top comments to include, not the join condition

+1

>     * having a `limit` clause was applying to the entire select of comments, not just the association as part of the include. As a result: only a max of two top comments **across all the articles** were being loaded 😱

Oooh! Didn't notice it in the review, weird though as there is a `WHERE commentable_id = article_id` condition in the final query. Anyway, you can still call `.top_comments.limit(2)` in the JSON if it gets slow :D 

LGTM!
